more(1) re-engineering resubmission

kerolasa opened the pull request:
This is a resubmission of https://github.com/karelzak/util-linux/pull/694 with a 'fix' to mac build that is marking more(1) to be linux only.

While looking again into this I chose to slightly improve 'more: replace siglongjmp() and signal() calls with signalfd()' commit. The improvement is removal of 'stop_searching' and fiddling with signal handler and adding more_poll() that can be called from both more_key_command() and search() functions. The rest of the change set is kept the same as some months ago when these changes were first submitted to review.

karelzak commented on text-utils/more.c:
```diff
@@ -1076,13 +1076,18 @@ static void sigwinch_handler(struct more_control *ctl)
 	va_list argp;
 	char *arg;
-	char **args;
+	char **args __attribute__((__cleanup__(free_args))) = NULL;;
```
This is the first place in util-linux we have automatic cleanup :-)

Not sure if this is the right way how to introduce it. Maybe it would be better to NOT do it in more(1) re-engineering, but create another PR and add some free()-like calls and initialization macros to xalloc.h? Maybe also discuss it on mailing list.

kerolasa replied:
Good point. While I think cleanup attribute should be taken in use this is not the best way to introduce it to project. I will drop the `avoid using alloca() function` commit from updated pull.

Secondly making a separate change request that will add cleanup attribute and related functions to xalloc.h is better way. That effort is obviously now backlogged, and delivery will happen with some delay. Personally I think moving to automatic cleanups in selected code paths would make code better.

karelzak commented on text-utils/more.c:
```diff
@@ -2014,88 +2072,17 @@ int main(int argc, char **argv)
-			ctl.context.line_num = ctl.context.row_num = 0L;
-		}
-		ctl.argv_position++;
+	for (/*nothing*/; ctl.argv_position < ctl.num_files; ctl.argv_position++) {
```
Maybe more readable:
```
while (ctl.argv_position < ctl.num_files) {
  ...
  ctl.argv_position++
}
```

karelzak commented:
Looks good, please remove the __cleanup__ attribute.

kerolasa commented:
Updated pull dropped `avoid using alloca() function` commit and changed 63f96f682f7ee1f78ff0dbf0925aaee6814f94e8 `argv_position++` within while() loop. The rest is kept the  same.

karelzak commented:
Merged, I guess some people will hate signalfd(), but ... ;-)

BTW, why we do not support page-up/down as alias to 'f' and 'b' keys?

kerolasa commented:
> some people will hate signalfd()

Some people complain about some things, other people complain about everything. Such is life.

> why we do not support page-up/down as alias to 'f' and 'b' keys?

As far I can tell there are no good reason why page-up/down and arrow up/down should not work. But in practice key command input reading right now is a bit of a mess. I started looking how to sort that out last weekend, but I cannot promise when that work is ready to be reviewed. To be honest I am not in hurry with more(1). Full modernisation will likely take time better counted in months than weeks.

karelzak commented:
Yes, more(1) is there for backward compatibility and minimalism, otherwise we have less(1) ;-)
